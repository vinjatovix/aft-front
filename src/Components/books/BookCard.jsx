import React from "react";
import PropTypes from "prop-types";
import { NavLink } from "react-router-dom";

import { ListItem } from "../common/ListItem";
import { NavigateButton } from "../ui/buttons/NavigateButton";
import { AdminButtons } from "../ui/admin/AdminButtons";

const DEFAULT_COVER_BOOK =
  "https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg";

export const BookCard = ({ data, isEditor, actions, token }) => (
  <div className="flex-wrapper column">
    <ul
      data-testid="book-detail"
      className="book"
      onClick={() => {
        actions.detail();
      }}
    >
      {<img src={data?.img || DEFAULT_COVER_BOOK} alt={data.name} />}
      <ListItem name="Obra" children={data.name} />
      <ListItem name="Autor" className="small" children={data.author} />
    </ul>

    <NavLink to={`/book/${data._id}/characters`}>
      <NavigateButton text="ðŸŽ­ Personajes" />
    </NavLink>

    <NavLink to={`/book/${data._id}/scenes`}>
      <NavigateButton text="ðŸŽ¬ Escenas" />
    </NavLink>

    {isEditor && <AdminButtons token={token} item={data} actions={actions} />}
  </div>
);

BookCard.propTypes = {
  data: PropTypes.shape({
    name: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    img: PropTypes.string,
    description: PropTypes.string.isRequired,
  }).isRequired,
  isEditor: PropTypes.bool,
  actions: PropTypes.shape({
    setDataDetail: PropTypes.func.isRequired,
    add: PropTypes.func.isRequired,
    close: PropTypes.func.isRequired,
    delete: PropTypes.func.isRequired,
    edit: PropTypes.func.isRequired,
    detail: PropTypes.func.isRequired,
    refresh: PropTypes.func.isRequired,
  }).isRequired,
  token: PropTypes.string.isRequired,
};

BookCard.defaultProps = {
  isEditor: false,
};
