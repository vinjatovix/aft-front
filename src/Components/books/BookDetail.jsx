import PropTypes from "prop-types";

export const BookDetail = ({ data }) => {
  const defaultCoverBook =
    "https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg";

  return (
    <div className="detail">
      <p>{data.name}</p>
      <p className="small">{data.author}</p>
      <img src={data.img || defaultCoverBook} alt="book" />
      <p>{data.description}</p>
    </div>
  );
};

BookDetail.propTypes = {
  data: PropTypes.shape({
    name: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    img: PropTypes.string,
    description: PropTypes.string.isRequired,
  }).isRequired,
};
