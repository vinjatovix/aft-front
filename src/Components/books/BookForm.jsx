import React from "react";
import PropTypes from "prop-types";

export const BookForm = ({
  handleSubmit,
  handleChange,
  formState,
  message,
}) => {
  return (
    <form onSubmit={handleSubmit}>
      <div className="form-group">
        <input
          className="form-control"
          type="text"
          data-testid="name"
          id="name"
          name="name"
          value={formState.name}
          onChange={handleChange}
          placeholder="Título"
          autoFocus
        />
      </div>
      <div className="form-group">
        <input
          className="form-control"
          type="text"
          data-testid="author"
          id="author"
          name="author"
          value={formState.author}
          onChange={handleChange}
          placeholder="Autor"
        />
      </div>
      <div className="form-group">
        <input
          className="form-control"
          type="text"
          data-testid="img"
          id="img"
          name="img"
          value={formState.img}
          onChange={handleChange}
          placeholder="Url imagen"
        />
      </div>
      <div className="form-group">
        <input
          className="form-control"
          type="text"
          data-testid="description"
          id="description"
          name="description"
          value={formState.description}
          onChange={handleChange}
          placeholder="Descripción"
        />
      </div>
      <button type="submit" className="bg-green">
        Enviar
      </button>
      {message && (
        <p data-testid="message" className={message.type}>
          {message.text}
        </p>
      )}
    </form>
  );
};

BookForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  formState: PropTypes.shape({
    name: PropTypes.string,
    author: PropTypes.string,
    img: PropTypes.string,
    description: PropTypes.string,
  }).isRequired,
  message: PropTypes.shape({
    text: PropTypes.string,
    type: PropTypes.string,
  }),
};

BookForm.defaultProps = {
  message: null,
};
