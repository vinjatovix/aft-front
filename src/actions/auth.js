import { ACTION_TYPES } from "../types/types";

export const loginRequestAction = () => ({
  type: ACTION_TYPES.AUTH_LOGIN_REQUEST,
});

export const loginSuccessAction = (auth) => ({
  type: ACTION_TYPES.AUTH_LOGIN_SUCCESS,
  payload: auth,
});

export const loginFailureAction = (error) => ({
  type: ACTION_TYPES.AUTH_LOGIN_FAILURE,
  payload: error,
});

export const logoutAction = () => ({
  type: ACTION_TYPES.AUTH_LOGOUT,
});
