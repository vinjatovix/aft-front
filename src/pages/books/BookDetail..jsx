import PropTypes from "prop-types";
import React, { useEffect, useState } from "react";
import { NavLink } from "react-router-dom";
import { fetchCharactersByBookId } from "../../http";

export const BookDetail = ({ auth, data }) => {
  const defaultCoverBook =
    "https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg";
  const [characters, setCharacters] = useState([]);
  const getCharacters = async () => {
    const res = await fetchCharactersByBookId(auth.token, data._id);
    const bookChars = await res.json();

    setCharacters(bookChars);
  };

  useEffect(() => {
    getCharacters();
  }, [data]);

  return (
    <div className="detail">
      {data.name && <p>{data.name}</p>}
      {data.author && <p className="small">{data.author}</p>}
      {<img src={data.img || defaultCoverBook} alt="book" />}
      {data.description && <p>{data.description}</p>}
      {characters.length > 0 && (
        <NavLink to={`/characters/book/${data._id}`}>
          <p>Ver Personajes</p>
        </NavLink>
      )}
    </div>
  );
};

BookDetail.propTypes = {
  auth: PropTypes.object.isRequired,
  data: PropTypes.object.isRequired,
};
