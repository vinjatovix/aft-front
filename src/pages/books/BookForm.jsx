import React from "react";

export const BookForm = ({
  handleSubmit,
  handleChange,
  formState,
  message,
}) => {
  return (
    <form onSubmit={handleSubmit}>
      <div className="form-group">
        <input
          className="form-control"
          type="text"
          id="name"
          name="name"
          value={formState.name}
          onChange={handleChange}
          placeholder="TÃ­tulo"
          autoFocus
        />
      </div>
      <div className="form-group">
        <input
          className="form-control"
          type="text"
          id="author"
          name="author"
          value={formState.author}
          onChange={handleChange}
          placeholder="Autor"
        />
      </div>
      <div className="form-group">
        <input
          className="form-control"
          type="text"
          id="img"
          name="img"
          value={formState.img}
          onChange={handleChange}
          placeholder="Url imagen"
        />
      </div>
      <button type="submit" className="bg-green">
        Enviar
      </button>
      {message && <p className={message.type}>{message.text}</p>}
    </form>
  );
};
