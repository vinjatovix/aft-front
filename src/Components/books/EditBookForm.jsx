import React, { useState } from "react";
import PropTypes from "prop-types";
import { BookForm } from "./BookForm";
import { useForm } from "../../hooks/useForm";
import { fetchAddBook, fetchUpdateBook } from "../../http";
import { resetMessage } from "../../helpers/resetMessage";

export const EditBookForm = ({ auth, data, actions }) => {
  const successText = data ? "Obra actualizada" : "Obra aÃ±adida";
  const initMessage = { type: null, text: null };
  const [message, setMessage] = useState(initMessage);
  const [formState, handleChange, resetForm] = useForm(data ? { ...data } : {});

  const submit = async () => {
    const res = data
      ? await fetchUpdateBook(auth.token, data._id, formState)
      : await fetchAddBook(auth.token, formState);

    if (res.ok) {
      setMessage({
        type: "success",
        text: successText,
      });
      resetMessage(setMessage, initMessage, resetForm);
      actions.refresh();
    } else {
      const error = await res.json();
      setMessage({ type: "error", text: error.errors });
      resetMessage(setMessage, initMessage, resetForm);
    }
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    submit();
  };

  return (
    <BookForm
      formState={formState}
      handleChange={handleChange}
      handleSubmit={handleSubmit}
      message={message}
    />
  );
};

EditBookForm.propTypes = {
  auth: PropTypes.object.isRequired,
  data: PropTypes.object,
  actions: PropTypes.shape({
    setDataDetail: PropTypes.func.isRequired,
    add: PropTypes.func.isRequired,
    close: PropTypes.func.isRequired,
    delete: PropTypes.func.isRequired,
    edit: PropTypes.func.isRequired,
    detail: PropTypes.func.isRequired,
    refresh: PropTypes.func.isRequired,
  }).isRequired,
};

EditBookForm.defaultProps = {
  data: null,
};
