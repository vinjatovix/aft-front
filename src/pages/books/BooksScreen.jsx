import React, { useContext, useEffect } from "react";
import { Content } from "../../Components/common/Content";
import { CardGrid } from "../../Components/common/CardGrid";
import { Modal } from "../../Components/common/modals/Modal";
import { ConfirmModal } from "../../Components/common/modals/ConfirmModal";
import { useModals } from "../../hooks/useModals";

import { useCanvas } from "../../hooks/useCanvas";
import { fetchBooks, fetchDeleteBook } from "../../http";
import { UserContext } from "../../contexts/UserContext";
import { BookDetail } from "./BookDetail";
import EditBookForm from "./EditBookForm";

export const BooksScreen = () => {
  const { auth } = useContext(UserContext);
  const { roles = [] } = auth.user || [];
  const isAdmin = roles.includes("aft.admin");

  const [modals, dispatchModals] = useModals();
  const [bookList, dispatchBookList] = useCanvas();

  const modalActions = {
    ...dispatchModals,
    refresh: () => dispatchBookList.setUpdate(),
  };

  const getBooks = async () => {
    const res = await fetchBooks(auth.token);
    if (res.ok) {
      const data = await res.json();
      dispatchBookList.setCanvas(data);
    }
  };

  const deleteBook = async (token, _id) => {
    const res = await fetchDeleteBook(token, _id);
    if (res.ok) {
      modalActions.close();
      dispatchBookList.setUpdate();
    }
  };

  useEffect(() => {
    auth.token && getBooks();
  }, [auth.token, bookList.updatedAt]);

  return (
    <>
      <Content modals={modals} modalActions={modalActions} isAdmin={isAdmin}>
        <h1>Obras: {dispatchBookList.count}</h1>
        {/* <input type="text" placeholder="Buscar..." /> */}

        <CardGrid
          type="book"
          data={bookList.data}
          isAdmin={isAdmin}
          actions={modalActions}
          token={auth.token}
        />
      </Content>

      {modals.blur && (
        <Modal actions={{ close: modalActions.close }}>
          {modals.detail && <BookDetail auth={auth} data={modals.data} />}
          {modals.edit && (
            <EditBookForm
              auth={auth}
              data={modals.data}
              actions={modalActions}
            />
          )}
          {modals.add && <EditBookForm auth={auth} actions={modalActions} />}
          {modals.delete && (
            <ConfirmModal
              auth={auth}
              data={modals.data}
              modalActions={modalActions}
              callback={deleteBook}
              message="¿Estás seguro de que quieres eliminar esta obra?"
            />
          )}
        </Modal>
      )}
    </>
  );
};
