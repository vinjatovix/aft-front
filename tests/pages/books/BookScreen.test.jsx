import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { BooksScreen } from "../../../src/pages/books/BooksScreen";
import actions from "../../fixtures/actions";
import books from "../../fixtures/books";

jest.mock("../../../src/helpers/getApiHost", () => ({
  getApiHost: jest.fn().mockReturnValue("host"),
}));

const auth = {
  token: "algo",
  user: {
    username: "user",
    roles: ["aft.user"],
  },
};

const modals = {
  add: false,
  blur: false,
  delete: false,
  detail: false,
  edit: false,
  data: {},
};

const Entity = {
  data: [books[0]],
  updatedAt: "",
};

const dispatchEntity = {
  count: Entity.data.length,
  data: [],
  updatedAt: "",
};

describe("BookScreen - page", () => {
  it("should render as user", () => {
    render(
      <Router>
        <BooksScreen
          auth={auth}
          dispatchEntity={dispatchEntity}
          modalActions={actions}
          modals={modals}
          Entity={Entity}
          isAdmin={false}
        />
      </Router>
    );

    const actionButtons = screen.queryAllByTestId("action-button");
    const navigateButtons = screen.getAllByTestId("navigate-button");

    expect(actionButtons).toHaveLength(0);
    expect(navigateButtons).toHaveLength(2);
    expect(navigateButtons[0].textContent).toBe("🎭 Personajes");
    expect(navigateButtons[1].textContent).toBe("🎬 Escenas");
  });

  it("should render as admin", () => {
    render(
      <Router>
        <BooksScreen
          auth={auth}
          dispatchEntity={dispatchEntity}
          modalActions={actions}
          modals={modals}
          Entity={Entity}
          isAdmin={true}
        />
      </Router>
    );
    const actionButtons = screen.getAllByTestId("action-button");
    const navigateButtons = screen.getAllByTestId("navigate-button");

    expect(actionButtons).toHaveLength(3);
    expect(actionButtons[0].innerHTML).toBe("Añadir");
    expect(actionButtons[1].innerHTML).toBe("📝 Editar");
    expect(actionButtons[2].innerHTML).toBe("🗑️ Eliminar");
    expect(navigateButtons).toHaveLength(2);
    expect(navigateButtons[0].textContent).toBe("🎭 Personajes");
    expect(navigateButtons[1].textContent).toBe("🎬 Escenas");
  });

  it("should render as admin with no books and add button", () => {
    render(
      <Router>
        <BooksScreen
          auth={auth}
          dispatchEntity={dispatchEntity}
          modalActions={actions}
          modals={modals}
          Entity={{ ...Entity, data: [] }}
          isAdmin={true}
        />
      </Router>
    );
    const actionButtons = screen.getAllByTestId("action-button");
    const navigateButtons = screen.queryAllByTestId("navigate-button");

    expect(actionButtons).toHaveLength(1);
    expect(navigateButtons).toHaveLength(0);
  });

  it("should show the number of books in the title", () => {
    const data = [...books];

    const { container } = render(
      <Router>
        <BooksScreen
          auth={auth}
          dispatchEntity={dispatchEntity}
          modalActions={actions}
          modals={modals}
          Entity={{ ...Entity, data }}
          isAdmin={false}
        />
      </Router>
    );
    const actionButtons = screen.queryAllByTestId("action-button");
    const title = screen.getByTestId("Obras:");

    expect(container).toMatchSnapshot();
    expect(actionButtons).toHaveLength(0);
    expect(title.textContent).toBe(`Obras: ${data.length}`);
  });
});
