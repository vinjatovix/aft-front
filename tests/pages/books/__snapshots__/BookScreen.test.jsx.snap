// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookScreen - page should show the number of books in the title 1`] = `
<div>
  <div
    class="content"
  >
    <div
      class="content_header"
    />
    <h3
      data-testid="Obras:"
    >
      Obras:
       
      5
    </h3>
    <div
      class="card-grid"
    >
      <div
        class="card"
      >
        <div
          class="flex-wrapper column"
        >
          <ul
            class="book"
            data-testid="book-detail"
          >
            <img
              alt="transmitter Refined"
              src="https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg"
            />
            <li
              class=""
              data-testid="Obra"
            >
              transmitter Refined
            </li>
            <li
              class="small"
              data-testid="Autor"
            >
              Theresa Larkin
            </li>
          </ul>
          <a
            class=""
            href="/characters/book/6301216103b6fbb712bbde7c"
          >
            <button
              class="bg-blue"
              data-testid="navigate-button"
            >
              ðŸŽ­ Personajes
            </button>
          </a>
          <a
            class=""
            href="/scenes/book/6301216103b6fbb712bbde7c"
          >
            <button
              class="bg-blue"
              data-testid="navigate-button"
            >
              ðŸŽ¬ Escenas
            </button>
          </a>
        </div>
      </div>
      <div
        class="card"
      >
        <div
          class="flex-wrapper column"
        >
          <ul
            class="book"
            data-testid="book-detail"
          >
            <img
              alt="alarm Tasty"
              src="https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg"
            />
            <li
              class=""
              data-testid="Obra"
            >
              alarm Tasty
            </li>
            <li
              class="small"
              data-testid="Autor"
            >
              Jay Hintz
            </li>
          </ul>
          <a
            class=""
            href="/characters/book/6301212c03b6fbb712bbde7a"
          >
            <button
              class="bg-blue"
              data-testid="navigate-button"
            >
              ðŸŽ­ Personajes
            </button>
          </a>
          <a
            class=""
            href="/scenes/book/6301212c03b6fbb712bbde7a"
          >
            <button
              class="bg-blue"
              data-testid="navigate-button"
            >
              ðŸŽ¬ Escenas
            </button>
          </a>
        </div>
      </div>
      <div
        class="card"
      >
        <div
          class="flex-wrapper column"
        >
          <ul
            class="book"
            data-testid="book-detail"
          >
            <img
              alt="Dr. microchip Small"
              src="https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg"
            />
            <li
              class=""
              data-testid="Obra"
            >
              Dr. microchip Small
            </li>
            <li
              class="small"
              data-testid="Autor"
            >
              Darrell Senger
            </li>
          </ul>
          <a
            class=""
            href="/characters/book/62ffb64f42dc23efb3c3c0c4"
          >
            <button
              class="bg-blue"
              data-testid="navigate-button"
            >
              ðŸŽ­ Personajes
            </button>
          </a>
          <a
            class=""
            href="/scenes/book/62ffb64f42dc23efb3c3c0c4"
          >
            <button
              class="bg-blue"
              data-testid="navigate-button"
            >
              ðŸŽ¬ Escenas
            </button>
          </a>
        </div>
      </div>
      <div
        class="card"
      >
        <div
          class="flex-wrapper column"
        >
          <ul
            class="book"
            data-testid="book-detail"
          >
            <img
              alt="Mr. panel Practical"
              src="https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg"
            />
            <li
              class=""
              data-testid="Obra"
            >
              Mr. panel Practical
            </li>
            <li
              class="small"
              data-testid="Autor"
            >
              Shaun Mohr
            </li>
          </ul>
          <a
            class=""
            href="/characters/book/62ffb63c42dc23efb3c3c0c2"
          >
            <button
              class="bg-blue"
              data-testid="navigate-button"
            >
              ðŸŽ­ Personajes
            </button>
          </a>
          <a
            class=""
            href="/scenes/book/62ffb63c42dc23efb3c3c0c2"
          >
            <button
              class="bg-blue"
              data-testid="navigate-button"
            >
              ðŸŽ¬ Escenas
            </button>
          </a>
        </div>
      </div>
      <div
        class="card"
      >
        <div
          class="flex-wrapper column"
        >
          <ul
            class="book"
            data-testid="book-detail"
          >
            <img
              alt="El principito"
              src="https://media.v2.siweb.es/uploaded_thumb_medium/bd4e5fefd47a03f444d7dce77040dff3/principito_con_letra.jpg"
            />
            <li
              class=""
              data-testid="Obra"
            >
              El principito
            </li>
            <li
              class="small"
              data-testid="Autor"
            >
              Saint-ExupÃ©ry
            </li>
          </ul>
          <a
            class=""
            href="/characters/book/62fed32c76d7b41305161d8a"
          >
            <button
              class="bg-blue"
              data-testid="navigate-button"
            >
              ðŸŽ­ Personajes
            </button>
          </a>
          <a
            class=""
            href="/scenes/book/62fed32c76d7b41305161d8a"
          >
            <button
              class="bg-blue"
              data-testid="navigate-button"
            >
              ðŸŽ¬ Escenas
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
